crates/accelerate/src/quantum_circuit/intern_context.rs

```diff
@@ -0,0 +1,71 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use hashbrown::HashMap;
+use pyo3::exceptions::PyRuntimeError;
+use pyo3::PyResult;
+use std::sync::Arc;
+
+pub type IndexType = u32;
+pub type BitType = u32;
+
+/// A Rust-only data structure (not a pyclass!) for interning
+/// `Vec<BitType>`.
+///
+/// Takes ownership of vectors given to [InternContext.intern]
+/// and returns an [IndexType] index that can be used to look up
+/// an _equivalent_ sequence by reference via [InternContext.lookup].
+#[derive(Clone, Debug)]
+pub struct InternContext {
+    slots: Vec<Arc<Vec<BitType>>>,
+    slot_lookup: HashMap<Arc<Vec<BitType>>, IndexType>,
+}
+
+impl InternContext {
+    pub fn new() -> Self {
+        InternContext {
+            slots: Vec::new(),
+            slot_lookup: HashMap::new(),
+        }
+    }
+
+    /// Takes `args` by reference and returns an index that can be used
+    /// to obtain a reference to an equivalent sequence of `BitType` by
+    /// calling [CircuitData.lookup].
+    pub fn intern(&mut self, args: Vec<BitType>) -> PyResult<IndexType> {
+        if let Some(slot_idx) = self.slot_lookup.get(&args) {
+            return Ok(*slot_idx);
+        }
+
+        let args = Arc::new(args);
+        let slot_idx: IndexType = self
+            .slots
+            .len()
+            .try_into()
+            .map_err(|_| PyRuntimeError::new_err("InternContext capacity exceeded!"))?;
```

Comment:
Technically, this isn't the error that gets raised in that case. I wasn't actually raising an error at all in those cases (which should have been in calls to `add_{qubit|clbit}`. This is fixed in cd3b920.

The specific error here occurs when there are `> 2^32` unique sequences of indices added to the interner.

In all cases, I'm not sure that a specific error type is useful, since these aren't recoverable errors. If people are hitting these, they'll need to rewrite their code or file a bug on us to add automatic scaling up to 64 bits.

Reply:
Yeah, that's fair `RuntimeError` is fine in this case if it's not expected to be recoverable. Although, do you think `OverflowError` is a better choice? https://docs.python.org/3/library/exceptions.html#OverflowError

I do think we should still improve the error message on this though, as an end user if I somehow exceeded 2^32 different qargs combinations and got `InternContext capacity exceeded!` I wouldn't know what that meant.

Reply:
IMO, `OverflowError` would make sense if it weren't an implementation detail that the limitation comes from our use of `u32` for interner IDs.

I'll improve this error message to include an explanation that too many unique bit lists are in play.